pkg/querier/queryrange/limits.go

```diff
@@ -178,6 +186,83 @@ func (l limitsMiddleware) Do(ctx context.Context, r queryrangebase.Request) (que
 	return l.next.Do(ctx, r)
 }
 
+type querySizeLimiter struct {
+	next              queryrangebase.Handler
+	maxQueryBytesRead func(string) int
+	errorTemplate     string
+}
+
+// NewQuerySizeLimiterMiddleware creates a new Middleware that enforces query size limits.
+// The errorTemplate should format two strings: the bytes that would be read and the bytes limit.
+func NewQuerySizeLimiterMiddleware(maxQueryBytesRead func(string) int, errorTemplate string) queryrangebase.Middleware {
```

Comment:
> Skip non-TSDB schemas since they don't have support for this.

Implemented in [1674e67](https://github.com/grafana/loki/pull/8670/commits/1674e67060515e7c747d30e28f261ad2cdf92757).

> We will want to split these first-pass IndexStats requests in a maximum of 24h chunks since our indices are sharded by 24h periods. Take a look at NewSeriesTripperware for an idea of how this works. Note: adding this limit may expose new bottlenecks in our idx-gws+tsdb that we need to address.

By "first-pass IndexStats requests" here: Do you mean the IndexStat requests to get the stats for all the matcher groups in the whole query (before splitting and sharding)? 

I'm not sure I understand the 24h part, do we have any docs so I can better understand this 24h sharding at the index level?